petab/core.py

```diff
@@ -38,87 +33,110 @@ class Problem:
+                 parameter_df: pd.DataFrame = None):
 
-        if model_name is None:
-            model_name = os.path.splitext(os.path.split(sbml_file)[-1])[0]
-        self.model_name = model_name
-
-        self.measurement_file = measurement_file
-        self.condition_file = condition_file
-        self.parameter_file = parameter_file
-        self.sbml_file = sbml_file
-
-        self.condition_df = None
-        self.measurement_df = None
-        self.parameter_df = None
-        self._load_dfs()
+        self.condition_df = condition_df
+        self.measurement_df = measurement_df
+        self.parameter_df = parameter_df
 
         self.sbml_reader = None
         self.sbml_document = None
-        self.sbml_model = None
-        self._load_sbml()
+        self.sbml_model = sbml_model
 
     def __getstate__(self):
         state = self.__dict__.copy()
-        # libsbml stuff cannot be serialized
-        # dfs can be recreated
-        for key in ['sbml_reader', 'sbml_document', 'sbml_model',
-                    'condition_df', 'measurement_df', 'parameter_df']:
+
+        # libsbml stuff cannot be serialized directly
+        if self.sbml_model:
+            sbml_document = self.sbml_model.getSBMLDocument()
+            sbml_writer = libsbml.SBMLWriter()
+            state['sbml_string'] = sbml_writer.writeSBMLToString(sbml_document)
+
+        exclude = ['sbml_reader', 'sbml_document', 'sbml_model']
+        for key in exclude:
             state.pop(key)
+
         return state
 
     def __setstate__(self, state):
+        # load SBML model from pickled string
+        sbml_string = state.pop('sbml_string', None)
+        if sbml_string:
+            self.sbml_reader = libsbml.SBMLReader()
+            self.sbml_document = \
+                self.sbml_reader.readSBMLFromString(sbml_string)
+            self.sbml_model = self.sbml_document.getModel()
+
         self.__dict__.update(state)
 
-        # load sbml from file name
-        self._load_sbml()
+    @staticmethod
+    def from_files(sbml_file: str = None,
+                   condition_file: str = None,
+                   measurement_file: str = None,
+                   parameter_file: str = None):
+        """
+        Factory method to load model and tables from files.
+
+        Arguments:
+            sbml_file: PEtab SBML model
+            condition_file: PEtab condition table
+            measurement_file: PEtab measurement table
+            parameter_file: PEtab parameter table
+        """
 
-        # reload dfs
-        self._load_dfs()
+        problem = Problem()
+        if condition_file:
+            problem.condition_df = get_condition_df(condition_file)
+        if measurement_file:
+            problem.measurement_df = get_measurement_df(measurement_file)
+        if parameter_file:
+            problem.parameter_df = get_parameter_df(parameter_file)
+        if sbml_file:
+            problem._load_sbml(sbml_file)
```

Comment:
in the sense of object orientation, one could first create the dfs and model, and then pass all to a new Problem(args=)